### PR TITLE
Provide a CharacterNotifications bus to drive association of active PhysX character and Network Character Controllers

### PR DESCRIPTION
**Pull request by mbalfour-amzn**
## What does this PR do?

The recent change to Static Rigid Bodies changed the timing of the Character Controller initialization. It used to occur during the component's Activate(), which meant that other components on the entity could be guaranteed that they could fetch a pointer to the component during their Activate() calls, as long as they had the service dependencies set up correctly.

However, the Character Controller now activates during OnEntityActivated(), which means that other components can no longer rely on the service dependencies to know when the Character Controller is available.

To fix this, the Character Controller now has a new notification bus that it uses to signal when the character is activated and deactivated. Dependent components can listen for these notifications to know when it is safe to communicate with the Character Controller and when it isn't.

## How was this PR tested?

Ran the Multiplayer Sample and verified that the character can once again move around in networked conditions.
Run physx gem unit tests.
Run multiplayer gem unit tests.

### CODE REVIEW

File: Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkCharacterComponent.h
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkCharacterComponent.h:59]: Should we have an OnCharacterDeactiveated() call?

Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkCharacterComponent.h:59]: Maybe on `OnPhysicsCharacterActivated()` to make it intent clearer?

File: Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp
Comment (lmbr-pip) [inline Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp:116]:
```
void NetworkCharacterComponent::OnPhysicsCharacterDeactivated()
{
    m_physicsCharacter = nullptr;
}
```

**Review by lmbr-pip — COMMENTED** [review]
LG.

Thanks for adding this.

File: Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp
Comment (moraaar) [inline Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp:120]: Notice how `NetworkCharacterComponent` is only using `CharacterNotificationBus` for "creation" only since it disconnects right away.

I did this to keep the exact same logic as before since it wasn't handling the character being enabled/disabled. But having into consideration that enabling/disabling a character recreates the body entirely internally my suggestion would be to keep listening to `OnCharacterActivated` and `OnCharacterDeactivated` to refresh everything inside `NetworkCharacterComponent`. The only reason I didn't doit is because I didn't know the consequences of such change and so I kept the same logic as before.

This is now exactly the same as in NetworkRigidBodyComponent and NetworkRigidBodyControllerComponent, so they are aligned now, but if you guys prefer to go ahead and change it please change it in all of them.

File: Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp
Comment (moraaar) [reply Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp]: I answered this in a separate code comment.

**Review by moraaar — COMMENTED** [review]
I'll leave other folks to approve since I contributed, but it's got my **virtual approval** :D

**Comment by moraaar** [comment]
> Not a fan of the spaghetti of messages being thrown around when something activates just to get all the timing down right. Any chance we could have the physics folks look at their change again and make sure the ordering change was actually necessary? It's a breaking change and we're not sure what else may have broken..

In this case the change was necessary, I talked today again with the other physics folks and we still think this is the appropriate change.

It is indeed an impactful change and we spent 1 week testing it with 4 people and PRs went sent to multiple sigs. It's difficult to cover everything, we tried to cover as many repos as possible, but unfortunately network character component fell through the cracks and I am very sorry that affected other members of the team. I'll be paying much more attention to it from now on.

Please do not hesitate to contact me with any questions about colliders or rigid bodies.

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (hultonha) [inline Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:108]: uber nit: I believe `//!` style is generally preferred now if possible

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (hultonha) [inline Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:113]: question: Based on the recent EBus changes by @lawsonamzn I wonder if it's preemptively worth moving this to the .cp and using the new `extern` approach? It might be worth waiting to profile the compile time I just wanted to sanity check in case.

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (hultonha) [inline Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:129]: uber nit: make the notifications -> send the notifications

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (hultonha) [inline Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:156]: nit (repeated): Prefer `//!` (the reason for this was for accessibility because by default in Visual Studio in the dark them the color selected for these 'XML' comments is really hard to read, there is no such problem with `//!` however so all new code should prefer it).

File: Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterControllerComponent.cpp
Comment (hultonha) [inline Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterControllerComponent.cpp:596]: nit: Could extract `EntityId` to a local to stop having to call this function repeatedly

```
const AZ::EntityId entityId = GetEntityId();
AZ::TransformNotificationBus::Handler::BusConnect(entityId);
... // etc
```

**Review by hultonha — APPROVED** [review]
Change looks good to me - I realize this change is needed ASAP but I think it would be great to take a breath after this change and think about how we could create a test to encapsulate this behavior (this could serve as an easy to find example as well to see how to handle this particular case). We just need some way to now 'lock' in this behavior so something breaks if we change something in RigidBody in future (this can be in a separate PR but it would be great to do and see if we can have a simplified example of how to handle the `OnEntityActivated` behavior)

File: Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp
Comment (moraaar) [reply Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp:120]: @lmbr-pip @kberg-amzn would be doing `m_physicsCharacter = nullptr;` in a `OnCharacterDeactivated` function enough to revert all that `OnCharacterActivated` is doing or would it need to do something else. Is it harmless that if the character is activate again all the code of `OnCharacterActivated` is executed again?

Comment (mbalfour-amzn) [reply Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp:120]: @lmbr-pip @kberg-amzn I'm not very familiar with this code, but I'm from looking at the OnCharacterActivated(), I'm thinking we need the following?
- Add the OnCharacterDeactivated() to set m_physicsCharacter to null.
- Move these lines to Activate() instead of OnCharacterActivated(), so that way in case OnCharacterActivated/Deactivated ever gets called more than once, things continue to work ok. These seem like they shouldn't get called more than once, and that they're ok to call even before the character exists?
```
        GetNetBindComponent()->AddEntitySyncRewindEventHandler(m_syncRewindHandler);
```
```
        if (!HasController())
        {
            GetNetworkTransformComponent()->TranslationAddEvent(m_translationEventHandler);
        }
```

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (greerdv) [inline Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:166]: I'm wondering if there's any value in having the entity Id as a parameter here. Wouldn't you already need to know the entity Id in order to have connected to the bus?

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (moraaar) [reply Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:108]: This entire file is using /// so I'd rather keep it consistent and not do unnecessary changes in other places now.

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (moraaar) [reply Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:129]: I'll change that. Thanks.

File: Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp
Comment (moraaar) [reply Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp:120]: I would not move the code to `Activate()` because those register to call the function `NetworkCharacterComponent::OnSyncRewind` which needs `m_physicsCharacter` anyway, so I think it's more correct to do it all `OnCharacterActivated`.

Probably `OnCharacterDeactivated` would need to register those events too. Right?

**Review by greerdv — APPROVED** [review]

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (moraaar) [reply Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:129]: I'll wait if there is another change... I already kicked of AR and probably not worth it to do it again for this. But I'll keep it in mind if I have to push again.

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (mbalfour-amzn) [reply Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:113]: I believe @lawsonamzn specifically said _not_ to preemptively use the extern approach, that it should only be used for highly-used EBuses? This one seems like it would be more sparsely used IMO...

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (moraaar) [reply Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:166]: Good question. This is useful when it's required to know if more that 1 character is ready, so it can be easily used with a Multihandler and therefore it's very useful to know which one is calling.

File: Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp
Comment (moraaar) [reply Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp:120]: By the way, I'd be inclined to leave this for a subsequent PR.

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (mbalfour-amzn) [reply Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:166]: Yeah, it should be "gettable" inside the OnCharacterActivated() via "CharacterNotificationBus::GetCurrentBusId()". That's also admittedly a little bit less obvious and readable. I'm personally pretty ambivalent on this one, but that's why I didn't initially include the entityId in the first version of this event. :)

File: Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkCharacterComponent.h
Comment (moraaar) [reply Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkCharacterComponent.h]: I answered this in a separate code comment below.

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h
Comment (greerdv) [reply Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:166]: Yeah, I don't think it's a big deal either way, I was just curious.

File: Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkCharacterComponent.h
Comment (mbalfour-amzn) [reply Gems/Multiplayer/Code/Include/Multiplayer/Components/NetworkCharacterComponent.h]: I personally like the suggestion, but at least for now this naming is in line with the other physics bus names, so I'd prefer not to deviate. I'll leave it to the physics folks to decide if they want to make it more verbose in a follow-up PR.

File: Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp
Comment (mbalfour-amzn) [reply Gems/Multiplayer/Code/Source/Components/NetworkCharacterComponent.cpp:120]: i discussed with @AMZN-Olex and he's in favor of moving those chunks to Activate(), and adding the Disconnect() calls to Deactivate(), so that the network handling is kept more obviously separated from the physics character handling. i'll put that up for review in a follow-up PR.

**Comment by mbalfour-amzn** [comment]
This passed AR and development is green, so I'll submit as-is. I'll put up a follow-up PR to clean up the Activate() / Deactivate() logic in NetworkCharacterComponent.

**Pull request merged**
